File: web/script/src/managers/SceneManager.ts
import Game from "..";
import Bitmap from "@/core/Bitmap";
import Input from "@/core/Input";

import Scene_Base from "@/scenes/Scene_Base";

class SceneManagerClass {
	private _scene: Scene_Base | null = null;
	private _nextScene: Scene_Base | null = null;
	private readonly _stack: Array<typeof Scene_Base> = [];

	private _exiting = false;
	private _smoothDeltaTime = 1;
	private _elapsedTime = 0;

	private _previousScene: Scene_Base | null = null;
	private _previousClass: typeof Scene_Base | null = null;

	private _backgroundBitmap: Bitmap | null = null;

	private initialize () {
		Game.setTickHandler(deltaTime => this.update(deltaTime));
	}
	private DetermineRepeatNumber (deltaTime: number): number {
		this._smoothDeltaTime *= 0.8;
		this._smoothDeltaTime += Math.min(deltaTime, 2) * 0.2;

		if (this._smoothDeltaTime >= 0.9) {
			this._elapsedTime = 0;
			return Math.round(this._smoothDeltaTime);
		} else {
			this._elapsedTime += deltaTime;
			if (this._elapsedTime >= 1) {
				this._elapsedTime -= 1;
				return 1;
			}
			return 0;
		}
	}
	private terminate () {
		window.close();
	}

	public run (sceneClass: typeof Scene_Base) {
		this.initialize();
		this.goto(sceneClass);
		Game.startGameLoop();
	}

	public update (deltaTime: number) {
		const n = this.DetermineRepeatNumber(deltaTime);
		for (let i = 1; i <= n; i++)
			this.updateMain();
	}
	public updateMain () {
		Game.frameCount++;

		Input.update();
		this.changeScene();
		this.updateScene();
	}
	public changeScene () {
		if (this.isSceneChanging() && !this.isCurrentSceneBusy()) {
			if (this._scene) {
				this._scene.terminate();
				this.onSceneTerminate();
			}

			this._scene = this._nextScene;
			this._nextScene = null;
			if (this._scene) {
				this._scene.create();
				this.onSceneCreate();
				this._scene.afterCreate();
			}

			if (this._exiting)
				this.terminate();
		}
	}
	public updateScene () {
		if (this._scene) {
			if (this._scene.isStarted())
				this._scene.update();
			else {
				this.onBeforeSceneStart();
				this._scene.start();
				this.onSceneStart();
				this._scene.afterStart();
			}
		}
	}

	public goto (sceneClass: typeof Scene_Base | null) {
		if (sceneClass)
			this._nextScene = new sceneClass();

		if (this._scene)
			this._scene.stop();
	}
	public push (sceneClass: typeof Scene_Base) {
		if (this._scene)
			this._stack.push((this._scene.constructor as typeof Scene_Base).Class);
		this.goto(sceneClass);
	}
	public pop () {
		if (this._stack.length > 0)
			this.goto(this._stack.pop()!);
		else
			this.exit();
	}
	public exit () {
		this.goto(null);
		this._exiting = true;
	}
	public clearStack () {
		this._stack.splice(0, this._stack.length);
	}
	public stop () {
		Game.stopGameLoop();
	}

	public snap () {
		return Bitmap.snap(this._scene!);
	}
	public snapForBackground () {
		if (this._backgroundBitmap)
			this._backgroundBitmap.destroy();
		this._backgroundBitmap = this.snap();
	}
	public backgroundBitmap () {
		return this._backgroundBitmap;
	}

	public resume () {
		Input.update();
		Game.startGameLoop();
	}

	public isGameActive () {
		try {
			return document.hasFocus();
		} catch (e) {
			return true;
		}
	};

	public onSceneTerminate () {
		this._previousScene = this._scene;
		this._previousClass = (this._scene!.constructor as typeof Scene_Base).Class;
		Game.setStage(null);
	};
	public onSceneCreate () {
		// Placeholder
	};
	public onBeforeSceneStart () {
		if (this._previousScene) {
			this._previousScene.destroy();
			this._previousScene = null;
		}
	};
	public onSceneStart () {
		Game.setStage(this._scene!);
	};
	public isSceneChanging () {
		return this._exiting || !!this._nextScene;
	};

	public isCurrentSceneBusy () {
		return this._scene && this._scene.isBusy();
	};

	public isNextScene (sceneClass: typeof Scene_Base) {
		return this._nextScene && this._nextScene.constructor === sceneClass;
	};

	public isPreviousScene (sceneClass: typeof Scene_Base) {
		return this._previousClass === sceneClass;
	};
}

const SceneManager = new SceneManagerClass();
export default SceneManager;

File: VN/VNData/lua/core/SceneManager.lua
local class = {}
class._scene = nil
class._nextScene = nil
class._stack = {}
class._exiting = false
class._previousScene = nil
class._previousClass = nil
class._backgroundBitmap = nil
class._smoothDeltaTime = 1
class._elapsedTime = 0

local Meta = {}
local function initialize(self)
    Game.SetTickHandler(function(deltaTime)
        self:Update(deltaTime) -- 프레임 핸들러
    end)
end

local function DetermineRepeatNumber(self, deltaTime)
    self._smoothDeltaTime = self._smoothDeltaTime * 0.8
    self._smoothDeltaTime = self._smoothDeltaTime + math.min(deltaTime, 2) * 0.2
    if self._smoothDeltaTime >= 0.9 then
        self._elapsedTime = 0
        return math.round(self._smoothDeltaTime)
    else
        self._elapsedTime = self._elapsedTime + deltaTime
        if self._elapsedTime >= 1 then
            self._elapsedTime = self._elapsedTime - 1
            return 1
        end
        return 0
    end
end

function class:Run(sceneClass)
    -- if call(function()
    initialize(self)
    self:GoTo(sceneClass)
    Game.StartGameLoop()
    -- end) then return end
end

function class:Update(deltaTime)
    -- if pcall(function()
    local n, i = DetermineRepeatNumber(self, deltaTime), 0
    for i = 1, n do self:UpdateMain() end
    -- end) then return end
end
function class:UpdateMain()
    Game.frameCount = Game.frameCount + 1
    Input.Update()
    self:ChangeScene()
    self:UpdateScene()
end

function class:ChangeScene()
    if self:isSceneChanging() and not self:isCurrentSceneBusy() then
        if self._scene then
            self._scene:Terminate()
            self:onSceneTerminate()
        end

        self._scene = self._nextScene
        self._nextScene = nil
        if self._scene then
            self._scene:Create()
            self:onSceneCreate()
        end

        if self._exiting then self:Terminate() end
    end
end
function class:UpdateScene()
    if self._scene ~= nil then
        if self._scene:isStarted() then
            self._scene:Update() --
        else
            self:onBeforeSceneStart()
            self._scene:Start()
            self:onSceneStart()
        end
    end
end

function class:GoTo(sceneClass)
    if sceneClass ~= nil then self._nextScene = sceneClass() end
    if self._scene ~= nil then self._scene:Stop() end
end

function class:Push(sceneClass)
    Array.push(self._stack, self._scene.__constructor)
    self:GoTo(sceneClass)
end
function class:Pop()
    if #self._stack > 0 then
        self.GoTo(Array.pop(self._stack))
    else
        self:Exit()
    end
end
function class:Exit()
    self:GoTo(nil)
    self._exiting = true
end
function class:ClearStack() self._stack = {} end
function class:Stop() Game.StopGameLoop() end

function class:Snap() return Image.LoadRaw(Game.Snap()) end
function class:SnapForBackground()
    if self._backgroundBitmap ~= nil then self._backgroundBitmap:Unload() end
    self._backgroundBitmap = self:Snap()
end
function class:BackgroundBitmap() return self._backgroundBitmap end
function class:Resume() Game.StartGameLoop() end

function class:onSceneCreate()
    -- 씬 로드 시작
end
function class:onSceneTerminate()
    self._previousScene = self._scene
    self._previousClass = self._scene.__constructor
    Game.SetStage(nil)
end
function class:onBeforeSceneStart()
    if self._previousScene then
        self._previousScene:Unload()
        self._previousScene = nil
    end
end
function class:onSceneStart()
    -- 로딩 종료
    Game.SetStage(self._scene)
end
function class:onUnload() AudioManager:StopAll() end

function class:isSceneChanging() return self._exiting or self._nextScene end
function class:isCurrentSceneBusy() return self._scene and self._scene.isBusy() end
function class:isNextScene(sceneClass)
    return self._nextScene and self._nextScene.__constructor == sceneClass
end
function class:isPreviousScene(sceneClass)
    return self._previousClass == sceneClass
end

SceneManager = class
